### PR TITLE
[RHCLOUD-46774] Integrate group-principal membership parity checker

### PR DESCRIPTION
**Pull request by dsova06**
## Summary

Integrates the existing `GroupPrincipalInventoryChecker` into the background Kessel parity check task (`run_kessel_parity_checks_in_worker`). This adds group-principal membership verification as the third parity check alongside workspace hierarchy and custom role permissions.

**JIRA**: [RHCLOUD-46774](https://issues.redhat.com/browse/RHCLOUD-46774)

### Task integration (`tasks.py`)
- Queries all groups with prefetched principals for each tenant
- Builds relation tuples per group, filters out principals without `user_id`
- Calls `check_relationships()` and aggregates pass/fail per group
- New stats: `total_groups_checked`, `total_group_principal_relations_checked`
- Per-tenant output: `groups_checked`, `group_principal_check_passed`, `group_results`
- Tenant pass/fail now requires all three checks (workspace + custom role + group-principal)

### Checker optimization (`inventory_api_check.py`)
- Refactored `GroupPrincipalInventoryChecker.check_relationships()` to open a single gRPC channel for all principal checks instead of one channel per principal

### Tests
- 6 unit tests for `GroupPrincipalInventoryChecker` (all exist, some missing, all missing, empty, single, null user_id)
- 8 task integration tests (pass, fail, no principals, null user_id, multiple groups, combined failures, exception handling, disabled flag)

## Test plan

- [x] Unit tests for checker class cover all edge cases
- [x] Task integration tests verify stats aggregation and tenant pass/fail logic
- [x] Existing parity task tests still pass (24 tests)
- [x] Full test suite passes (3349 tests)
- [x] Linter and black formatting clean

[RHCLOUD-46774]: https://redhat.atlassian.net/browse/RHCLOUD-46774?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Integrate group–principal membership parity checking into the Kessel background parity task and extend reporting to include group-level results and statistics.

Enhancements:
- Extend the parity worker task to evaluate group–principal relationships per tenant alongside existing workspace and custom role checks, updating tenant pass/fail criteria and aggregated stats.
- Optimize GroupPrincipalInventoryChecker to reuse a single gRPC channel and stub for multiple relationship checks and handle empty relationship sets efficiently.

Tests:
- Add unit test suite for GroupPrincipalInventoryChecker covering empty, partial, and fully missing relationships, single principal, and principals without user IDs.
- Expand parity task tests to cover group–principal scenarios, including passing and failing tenants, groups without principals, principals lacking user IDs, multiple groups, checker exceptions, and ensuring the task short-circuits cleanly when disabled.

### CODE REVIEW

**Comment by dsova06** [comment]
@sourcery-ai review

**Review by dsova06 — COMMENTED** [review]
### Would be good to clean up

1. **One gRPC channel opened per group in the task loop.** Each call to `group_principal_checker.check_relationships(relationships)` inside the per-group loop opens and closes a gRPC channel via `create_client_channel_inventory`. For tenants with many groups, this means N separate connections. The optimization inside `check_relationships` itself (one channel for all principals within a group) is a good improvement over the original per-principal approach, but the task-level per-group pattern leaves room for a follow-up optimization -- e.g., passing a pre-opened channel/stub into the checker. That said, this is consistent with how `CustomRolePermissionChecker` works today (one channel per role via `check_inventory_core`), so it's not worse than existing patterns.

### Something to think about

1. **`check_relationships` implicitly assumes single-group input.** The `group_uuid` field in the return dict is overwritten on each loop iteration (`inventory_relation_assignments["group_uuid"] = r.resource.id`). The caller in `tasks.py` correctly passes only relationships for one group at a time, so the result is always correct. But the method signature accepts any list of relationships with no indication that mixing groups would produce a misleading `group_uuid`. A one-line docstring note like "relationships must belong to the same group" would make this explicit for future callers.

2. **`principal_count` counts checked principals, not total principals.** The field reports `len(relationships)` after filtering out None (principals without `user_id`). This is the right behavior for parity checking, but someone debugging a discrepancy between RBAC's principal count and the parity output might be briefly confused. Worth keeping in mind if this surfaces in operational use.

---

The code follows the project's established patterns well -- pre-try variable initialization for safe exception reporting, tenant-level per-item error isolation, graceful skip on empty input, and structured stats with per-entity detail. Test coverage is thorough (14 new tests covering all the important paths). The decision to keep the group-checking logic inline rather than extracting helpers, maintaining consistency with the workspace and custom role phases, is a reasonable call documented in the commit message.

Looks good to merge.

File: rbac/management/inventory_checker/inventory_api_check.py [hunk 1]
```diff
@@ -97,17 +97,25 @@ class GroupPrincipalInventoryChecker(InventoryApiBaseChecker):
...
-                    f"Relation missing: User ID {r.subject.subject.id} is not associated with Group ID {r.resource.id}"
+        if not relationships:
+            return inventory_relation_assignments
+
+        check_pairs = [(r, relation_tuple_to_check_request(r)) for r in relationships]
+
+        with create_client_channel_inventory(settings.INVENTORY_API_SERVER) as channel:
+            stub = inventory_service_pb2_grpc.KesselInventoryServiceStub(channel)
+            for r, check_request in check_pairs:
+                response = stub.Check(check_request)
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/inventory_checker/inventory_api_check.py]: Already addressed in 39fd551 -- check_pairs removed, requests constructed lazily inside the loop.

File: tests/management/inventory_checker/test_group_principal_checker.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    @patch("management.inventory_checker.inventory_api_check.create_client_channel_inventory")
+    @patch("management.inventory_checker.inventory_api_check.json_format.MessageToDict")
+    def test_check_relationships_all_exist(self, mock_message_to_dict, mock_create_channel):
+        """Test that check returns all relation_exists=True when all relations exist."""
+        self.group.principals.add(self.principal1, self.principal2, self.principal3)
+
+        mock_response = MagicMock(spec=CheckResponse)
+        mock_stub = self._setup_inventory_mocks(mock_create_channel, mock_response)
+        mock_message_to_dict.return_value = {"allowed": "ALLOWED_TRUE"}
+
```
[the remark this answers is not in the document]
Comment (dsova06) [reply tests/management/inventory_checker/test_group_principal_checker.py:84]: Already addressed in 39fd551 -- added assert_called_once() for both create_client_channel_inventory and KesselInventoryServiceStub.

File: rbac/management/tasks.py [hunk 3]
```diff
@@ -324,8 +331,43 @@ def run_kessel_parity_checks_in_worker():
             if role_results:
                 logger.info(f"Checked {len(role_results)} custom role(s) for tenant {org_id}")
 
-            # Tenant passes only if BOTH workspace and custom role checks pass
-            tenant_passed = workspace_check_passed and custom_role_check_passed
+            groups = Group.objects.filter(tenant=tenant).prefetch_related("principals")
```
[the remark this answers is not in the document]
Comment (dsova06) [reply rbac/management/tasks.py:430]: Intentionally keeping inline for now -- extracting only the group check while leaving workspace and custom role checks inline would be inconsistent. Will extract all phases together when a fourth check is added.

File: tests/management/inventory_checker/test_group_principal_checker.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    def _setup_inventory_mocks(self, mock_create_channel, mock_stub_responses):
+        """Helper to set up inventory API mocks."""
+        mock_stub = MagicMock()
+        if isinstance(mock_stub_responses, list):
+            mock_stub.Check.side_effect = mock_stub_responses
+        else:
+            mock_stub.Check.return_value = mock_stub_responses
+
+        mock_channel = MagicMock()
+        mock_channel.__enter__.return_value = mock_channel
```
[the remark this answers is not in the document]
Comment (dsova06) [reply tests/management/inventory_checker/test_group_principal_checker.py:69]: The existing assert_called_once() assertions for both channel and stub are sufficient to verify the single-channel optimization. Adding assert_called_once_with and refactoring the helper signature across all tests would be disproportionate churn for a marginal gain.

**Comment by dsova06** [comment]
@sourcery-ai resolve

**Comment by lpichler** [comment]
## Code Review Findings

### Issue 1: Channel management bypasses base class pattern (Design, Medium) (Info)

**Location:** `rbac/management/inventory_checker/inventory_api_check.py`, lines 97-120

The refactored `GroupPrincipalInventoryChecker.check_relationships()` opens its own gRPC channel with `create_client_channel_inventory()` and directly creates a stub, bypassing the base class `check_inventory_core()` method. Every other checker subclass (`WorkspaceRelationInventoryChecker`, `RoleRelationInventoryChecker`, `RoleBindingInventoryChecker`, `CustomRolePermissionChecker`) delegates to `check_inventory_core()` which handles channel lifecycle internally.

The stated motivation appears to be channel reuse within a single group's checks (opening one channel for all principals in a group rather than one per principal). However, the old code called `check_inventory_core()` per-principal which already opens/closes a channel per call -- the correct fix would be to batch all check requests and pass them as a list to `check_inventory_core()`, which already supports `List[CheckRequest]`.

**Suggested fix:** Build a list of `CheckRequest` objects, call `self.check_inventory_core(checks)` once, then iterate the per-request results. This would achieve the same channel-reuse benefit while preserving the base class pattern. If per-request results are needed (to identify which specific principal failed), the base class `check_inventory_core` would need a small enhancement to return per-request results rather than a single boolean, or a new method could be added.

---

### Issue 2: Redundant re-initialization of variables (Nit, Low) (Info)

**Location:** `rbac/management/tasks.py`, around lines 282-285 and 340-343

`group_principal_check_passed` and `group_results` are initialized before the `try` block (lines 282-283) and then re-initialized again inside the `try` block (lines 340-343). The second initialization is redundant since the variables already have the same values.

---

### 2. [Low] No explicit exception handling around group-principal check block (Info)

**File**: `rbac/management/tasks.py`

The group-principal checking code is placed inside the existing broad `try/except Exception` block, which catches all exceptions. While this means gRPC failures are handled, an exception during group iteration (e.g., a gRPC timeout on `check_relationships()` for group 2 of 5) will abort processing of all remaining groups for that tenant and mark the entire tenant as failed -- even though partial results exist.

This behavior is **consistent** with how workspace and custom role checks work in the same function (they also don't have per-item exception handling), so this is not a bug but rather a design choice that could be improved in a follow-up.

**Severity**: Low -- consistent with existing behavior.

**Comment by dsova06** [comment]
@sourcery-ai resolve

**Comment by dsova06** [comment]
@lpichler -- addressed your findings:

**Issue 1 (channel management):** Fixed. Added `_check_inventory_batch` to `InventoryApiBaseChecker` that returns per-request `list[bool]` results via a single gRPC channel. `check_relationships` now delegates to this base class method instead of managing its own channel/stub. Channel lifecycle is back in the base class where it belongs.

**Issue 2 (redundant re-init):** Fixed. Removed the redundant re-initialization of `group_principal_check_passed` and `group_results` inside the try block -- the pre-try initialization is sufficient.

**Issue 3 (per-item exception handling):** Acknowledged as consistent with existing workspace/custom role behavior. Will address across all phases together if we add per-item isolation in a follow-up.

**Comment by lpichler** [comment]
@dsova06 this needs rebase

**Review by lpichler — APPROVED** [review]
